app/com/arpnetworking/kairos/client/models/MetricsQuery.java

```diff
@@ -0,0 +1,547 @@
+            /**
+             * Creates a builder from a {@link Metric}.
+             *
+             * @param metric metric to use as the base
+             * @return a new Builder set with the fields of the metric argument
+             */
+            public static Builder from(final Metric metric) {
+                return new Builder().setTags(LinkedHashMultimap.create(metric._tags))
+                        .setAggregators(Lists.newArrayList(metric._aggregators))
+                        .setGroupBy(Lists.newArrayList(metric._groupBy))
+                        .setName(metric._name);
+            }
+
+            /**
+             * Public constructor.
+             */
+            public Builder() {
+                super(Metric::new);
+            }
+
+            /**
+             * Sets the name of the metric. Required. Cannot be null or empty.
+             *
+             * @param value the name of the metric
+             * @return this {@link Builder}
+             */
+            public Builder setName(final String value) {
+                _name = value;
+                return this;
+            }
+
+            /**
+             * Sets the tags. Optional. Cannot be null.
+             *
+             * @param value the tags
+             * @return this {@link Builder}
+             */
+            public Builder setTags(final Multimap<String, String> value) {
+                _tags.clear();
+                _tags.putAll(value);
+                return this;
+            }
+
+            /**
+             * Sets the group by. Optional. Cannot be null.
+             *
+             * @param value the group by clauses
+             * @return this {@link Builder}
+             */
+            public Builder setGroupBy(final List<GroupBy> value) {
+                _groupBy.clear();
+                _groupBy.addAll(value);
+                return this;
+            }
+
+            /**
+             * Sets the aggregators. Optional. Cannot be null.
+             *
+             * @param value the aggregators
+             * @return this {@link Builder}
+             */
+            public Builder setAggregators(final List<Aggregator> value) {
+                _aggregators.clear();
+                _aggregators.addAll(value);
+                return this;
+            }
+
+            /**
+             * Add an aggregator. Optional. Cannot be null
+             *
+             * @param value the aggregator to add
+             * @return this {@link Builder}
+             */
+            public Builder addAggregator(final Aggregator value) {
+                _aggregators.add(value);
+                return this;
+            }
+
+            /**
+             * Add a group by. Optional. Cannot be null
+             *
+             * @param value the aggregator to add
+             * @return this {@link Builder}
+             */
+            public Builder addGroupBy(final GroupBy value) {
```

Comment:
Not sure I see the value in the add and set, but need to read the rest of the code still.

Reply:
One problem I see:

- Call 1: setX
- Call 2: addX

Now it modifies the collection provided by caller 1. That's probably not cool. So I would recommend setting with ImmutableCollections, adding to ImmutableCollection builders and clearly the other as needed. Then override build and if you have a collection builder build it into the collection.

Reply:
I went straight with ImmutableX structures.  If you need addX, you should handle that yourself. I don't see the need to add the complexity to the builder at this time.